File: src/use-cases/game/PlayCards.js
/**
 * PlayCards Use Case
 * Handles playing cards from a player's hand
 */

const CardAnalyzer = require('../../infrastructure/bot/CardAnalyzer');
const logger = require('../../../logger');

class PlayCards {
    /**
     * @param {IPartyRepository} partyRepository - Party repository
     * @param {IUserRepository} userRepository - User repository
     */
    constructor(partyRepository, userRepository) {
        this.partyRepository = partyRepository;
        this.userRepository = userRepository;
    }

    /**
     * Execute the use case
     * @param {Object} request - Play cards request
     * @param {string} request.userId - User ID
     * @param {string} request.partyId - Party ID
     * @param {number[]} request.cardIds - Array of card IDs to play
     * @returns {Promise<Object>} Play result with updated game state
     */
    async execute({ userId, partyId, cardIds }) {
        try {
            // Validate input
            if (!userId || typeof userId !== 'string') {
                throw new Error('User ID is required');
            }

            if (!partyId || typeof partyId !== 'string') {
                throw new Error('Party ID is required');
            }

            if (!Array.isArray(cardIds) || cardIds.length === 0) {
                throw new Error('Card IDs are required');
            }

            // Verify user exists
            const user = await this.userRepository.findById(userId);
            if (!user) {
                throw new Error('User not found');
            }

            // Find party
            const party = await this.partyRepository.findById(partyId);
            if (!party) {
                throw new Error('Party not found');
            }

            // Check party is playing
            if (party.status !== 'playing') {
                throw new Error('Party is not in playing state');
            }

            // Get current round
            if (!party.currentRoundId) {
                throw new Error('No active round');
            }

            const round = await this.partyRepository.getRoundById(party.currentRoundId);
            if (!round) {
                throw new Error('Round not found');
            }

            // Get game state
            const gameState = await this.partyRepository.getGameState(partyId);
            if (!gameState) {
                throw new Error('Game state not found');
            }

            // Get player
            const players = await this.partyRepository.getPartyPlayers(partyId);
            const player = players.find(p => p.userId === userId);
            if (!player) {
                throw new Error('User is not in this party');
            }

            // Check if it's player's turn
            if (gameState.currentTurn !== player.playerIndex) {
                throw new Error('Not your turn');
            }

            // Check current action is PLAY
            if (gameState.currentAction !== 'play') {
                throw new Error('Current action is not PLAY');
            }

            // Check if player is eliminated
            const eliminatedPlayers = gameState.eliminatedPlayers || [];
            if (eliminatedPlayers.includes(player.playerIndex)) {
                throw new Error('Player is eliminated and cannot play');
            }

            // Get player's hand
            const playerHand = gameState.hands[player.playerIndex];
            if (!playerHand) {
                throw new Error('Player hand not found');
            }

            // Verify all cards are in player's hand
            for (const cardId of cardIds) {
                if (!playerHand.includes(cardId)) {
                    throw new Error(`Card ${cardId} not in hand`);
                }
            }

            // Validate card play (basic validation - could use utils.check_play from existing code)
            this.validateCardPlay(cardIds);

            // Remove cards from hand
            const newHand = playerHand.filter(id => !cardIds.includes(id));

            // Update game state
            const newHands = { ...gameState.hands };
            newHands[player.playerIndex] = newHand;

            // Add old lastCardsPlayed to discard pile (they're no longer pickable)
            const newDiscardPile = [...(gameState.discardPile || []), ...gameState.lastCardsPlayed];

            const newGameState = gameState.withUpdates({
                hands: newHands,
                cardsPlayed: cardIds,
                lastCardsPlayed: gameState.cardsPlayed,
                discardPile: newDiscardPile,
                currentAction: 'draw',
                lastAction: {
                    type: 'play',
                    playerIndex: player.playerIndex,
                    cardIds: cardIds,
                    timestamp: Date.now()
                }
            });

            // Save game state
            await this.partyRepository.saveGameState(partyId, newGameState);

            logger.info('Cards played', {
                userId: userId,
                username: user.username,
                partyId: partyId,
                roundId: round.id,
                cardIds: cardIds,
                cardsCount: cardIds.length
            });

            return {
                success: true,
                gameState: newGameState.toPublicObject(),
                cardsPlayed: cardIds,
                remainingCards: newHand.length
            };
        } catch (error) {
            logger.error('Play cards error', {
                userId,
                partyId,
                cardIds,
                error: error.message
            });

            throw error;
        }
    }

    /**
     * Validate card play
     * @param {number[]} cardIds - Card IDs to validate
     * @private
     */
    validateCardPlay(cardIds) {
        // Validate using CardAnalyzer
        if (!CardAnalyzer.isValidPlay(cardIds)) {
            throw new Error(
                'Invalid card play. Cards must be:\n' +
                '- Single card, OR\n' +
                '- 2+ cards of same rank (with optional jokers), OR\n' +
                '- 3+ cards in sequence of same suit (with optional jokers)'
            );
        }
    }
}

module.exports = PlayCards;

File: zapzap-rust/src/application/game/play_cards.rs
use std::sync::Arc;

use crate::domain::entities::PartyStatus;
use crate::domain::repositories::{PartyRepository, RepositoryError};
use crate::domain::services::execute_play;
use crate::domain::value_objects::GameAction;
use crate::infrastructure::bot::card_analyzer;

/// Play cards input
pub struct PlayCardsInput {
    pub party_id: String,
    pub user_id: String,
    pub card_ids: Vec<u8>,
}

/// Play cards output
pub struct PlayCardsOutput {
    pub cards_played: Vec<u8>,
    pub remaining_cards: usize,
}

/// Play cards use case
pub struct PlayCards<P: PartyRepository> {
    party_repo: Arc<P>,
}

impl<P: PartyRepository> PlayCards<P> {
    pub fn new(party_repo: Arc<P>) -> Self {
        Self { party_repo }
    }

    pub async fn execute(&self, input: PlayCardsInput) -> Result<PlayCardsOutput, PlayCardsError> {
        // Find party
        let party = self
            .party_repo
            .find_by_id(&input.party_id)
            .await?
            .ok_or(PlayCardsError::PartyNotFound)?;

        // Check party is playing
        if party.status != PartyStatus::Playing {
            return Err(PlayCardsError::PartyNotPlaying);
        }

        // Get player index
        let player_index = self
            .party_repo
            .get_player_index(&input.party_id, &input.user_id)
            .await?
            .ok_or(PlayCardsError::NotInParty)?;

        // Get game state
        let mut game_state = self
            .party_repo
            .get_game_state(&input.party_id)
            .await?
            .ok_or(PlayCardsError::NoGameState)?;

        // Check it's player's turn
        if game_state.current_turn != player_index {
            return Err(PlayCardsError::NotYourTurn);
        }

        // Check action is Play
        if game_state.current_action != GameAction::Play {
            return Err(PlayCardsError::WrongAction);
        }

        // Validate cards
        if input.card_ids.is_empty() {
            return Err(PlayCardsError::NoCardsSelected);
        }

        if !card_analyzer::is_valid_play(&input.card_ids) {
            return Err(PlayCardsError::InvalidCombination);
        }

        // Execute play
        execute_play(&mut game_state, &input.card_ids)
            .map_err(|e| PlayCardsError::GameError(e.to_string()))?;

        // Save game state
        self.party_repo
            .save_game_state(&input.party_id, &game_state)
            .await?;

        // Update round
        if let Some(mut round) = self.party_repo.get_current_round(&input.party_id).await? {
            round.current_turn = game_state.current_turn;
            round.current_action = game_state.current_action.as_str().to_string();
            self.party_repo.save_round(&round).await?;
        }

        let remaining_cards = game_state.get_hand(player_index).len();

        Ok(PlayCardsOutput {
            cards_played: input.card_ids,
            remaining_cards,
        })
    }
}

#[derive(Debug, thiserror::Error)]
pub enum PlayCardsError {
    #[error("Party not found")]
    PartyNotFound,
    #[error("Party is not playing")]
    PartyNotPlaying,
    #[error("Not in party")]
    NotInParty,
    #[error("No game state")]
    NoGameState,
    #[error("Not your turn")]
    NotYourTurn,
    #[error("Wrong action phase")]
    WrongAction,
    #[error("No cards selected")]
    NoCardsSelected,
    #[error("Invalid card combination")]
    InvalidCombination,
    #[error("Game error: {0}")]
    GameError(String),
    #[error("Repository error: {0}")]
    Repository(#[from] RepositoryError),
}
